### DISCUSSION TITLE
Using Xonsh and Rye together?

### DISCUSSION THREAD

**Discussion by jsquaredosquared** [xonsh/xonsh#5290] [Q&A] [answered]
Does anyone use Xonsh with [Rye](https://rye-up.com/)?

If I try installing Xonsh as a global tool with `rye install 'xonsh[full]'`, I can get the basic functionality of Xonsh. However, because Rye has a philosophy of not exposing `pip`, this prevents `xpip` from working and I can't install xontribs which I rely on.

Rye has the option of using a different tool to manage the global Python version. I tried using Pyenv, and this option works and I can install xontribs. However, using Xonsh inside a Rye-managed project requires me to first cd out of the project, start xonsh, then cd back in, which can be a bit annoying.

I have tried reading [Install xonsh with package and environment management system](https://github.com/anki-code/xonsh-cheatsheet?tab=readme-ov-file#install-xonsh-with-package-and-environment-management-system), but I'm still a bit confused.

Just curious if anyone else has tried using rye and xonsh together.

**Comment by anki-code** [comment]
Hey @jsquaredosquared! Thank you for using xonsh in new environments.
My thoughts:
1. It will be cool to have [copy-pastable list of the commands](https://github.com/anki-code/anki-code/tree/main?tab=readme-ov-file#my-principles) to repeat your experience.
2. If `xpip` provides different experience give us the example of how it should work and maybe you can create a xontrib using [xontrib-template](https://github.com/xonsh/xontrib-template) to support rye. 

(I see that my messages has "Maintainer" label but the "maintainer" is the team during a couple of years. Don't treat my comment as a comment from owner.)

**Comment by jsquaredosquared** [comment]
Thanks for the quick response :smiley: .

---
### 1. Can't install xontribs when using xonsh installed with rye package manager

After installing rye with `curl -sSf https://rye-up.com/get | bash
` (and making sure to allow rye to manage the global Python version during the setup process):
```shell
rye install 'xonsh[full]'
xonsh
```
Xontribs cannot be installed because of the following error:
```shell
xpip install xontrib-jupyter
```
`~/.rye/tools/xonsh/bin/python: No module named pip`

According to the [Rye philosphy](https://rye-up.com/philosophy/) (fourth bullet point): 

> No Pip: Rye uses pip, but it does not expose it. 


### 2. Can't start xonsh session if current directory is managed by rye but xonsh is installed globally or managed by another package manager

If you make a directory a Rye-managed project by running `rye init`, you can't start xonsh when you are in that directory because Rye will try to use the Python version from the virtual environment associated with that directory instead of the global Python version where xonsh was installed. It will throw the following error:
`pyenv: version '3.11.7' is not installed (set by <rye_project_dir>/.python-version)`

You would have to start Xonsh outside that directory before you can have access to it inside.

---

### Desired behaviour
Ideally I would like to use Rye to manage Xonsh instead of having to go through something like pipx or pyenv. I wish I had the knowledge to create a xontrib to fix this, but I wouldn't know how to begin :smiling_face_with_tear: .

**Reply by anki-code** [reply, marked as answer]
This works for me for case 1:
```xsh
docker run -it --rm python:3.12-slim bash
apt update && apt install -y curl
curl -sSf https://rye-up.com/get | bash 
source "$HOME/.rye/env"
rye install 'xonsh[full]'
xonsh

# Install pip using xonsh's python
# from xonsh shebang:
@($(head -n1 @$(which xonsh))[2:-1]) -m ensurepip --upgrade
# OR from xpip alias in case python already detected:
@(aliases['xpip'][0]) -m ensurepip --upgrade

# Now it works
xpip install xontrib-prompt-bar numpy
xontrib load prompt_bar
import numpy
```
I had a short look but can't find how to install packages into the appropriate tool environment using rye.

**Reply by anki-code** [reply]
For case 2. I cant repeat this:
> If you make a directory a Rye-managed project by running rye init, you can't start xonsh when you are in that directory

**Reply by jsquaredosquared** [reply]
Thanks! Now I don't have to give up my beloved Xonsh :shell: :smiley:

**Comment by jsquaredosquared** [comment]
Looks like the fix for case 1 solves case 2 also. Thanks :)
